### PR TITLE
Shanghai Update

### PR DESCRIPTION
**Pull request by pldespaigne**
Update the Yellow Paper from the _Paris_ hard-fork to the _Shanghai_ hard-fork.

- [x] EIP-3651: Warm COINBASE
- [x] EIP-3855: PUSH0 instruction
- [x] EIP-3860: Limit and meter initcode
- [x] EIP-4895: Beacon chain push withdrawals as operations
- [x] EIP-6049: Deprecate SELFDESTRUCT
- [x] General Updates

[Here](https://github.com/user-attachments/files/15835982/Paper.pdf) is the current pdf version of the Yellow Paper for this PR.

# EIP-3651: Warm COINBASE
- [specs](https://eips.ethereum.org/EIPS/eip-3651)
- [commit changes](https://github.com/ethereum/yellowpaper/pull/906/commits/6fb1032933785381ad4859bfb4be0400e4c45fe5)
- [review](https://github.com/ethereum/yellowpaper/pull/906#discussion_r1606402480) → [change: 8af40f2](https://github.com/ethereum/yellowpaper/pull/906/commits/8af40f2457dab040ddb191139579335ec95a53df)

<details>
<summary>View Screenshot Diff</summary>

|Before|After|
|-|-|
|3651_before|3651_after|

</details>

# EIP-3855: PUSH0 instruction
- [specs](https://eips.ethereum.org/EIPS/eip-3855)
- [commit changes](https://github.com/ethereum/yellowpaper/pull/906/commits/d997926306d953aaf12b8130ea2ce9b97c91e0a0)
- [review](https://github.com/ethereum/yellowpaper/pull/906#discussion_r1608345854)

<details>
<summary>View Screenshot Diff</summary>

|Before|After|
|-|-|
|3855_0_before|3855_0_after|
|3855_1_before|3855_1_after|

</details>

# EIP-3860: Limit and meter initcode
- [specs](https://eips.ethereum.org/EIPS/eip-3860)
- [commit changes](https://github.com/ethereum/yellowpaper/pull/906/commits/8428ca552b587abed700aced1d2d96d9ff22104a)
- [review](https://github.com/ethereum/yellowpaper/pull/906#discussion_r1610052660) → [changes: 263171f](https://github.com/ethereum/yellowpaper/pull/906/commits/263171f3e9b7599475e9f0abc8efe4ed1b5040e9)

<details>
<summary>View Screenshot Diff</summary>

|Before|After|
|-|-|
|3860_0_before|3860_0_after|
|3860_1_before|3860_1_after_1|
|3860_2_before|3860_2_after_1|
|3860_3_before|3860_3_after_1|
|3860_4_before|3860_4_after|

</details>

# EIP-4895: Beacon chain push withdrawals as operations
- [specs](https://eips.ethereum.org/EIPS/eip-4895)
- [commit changes](https://github.com/ethereum/yellowpaper/pull/906/commits/5173cdd43e7185b5e62c6816120273fe53039e62)
- [review](https://github.com/ethereum/yellowpaper/pull/906#discussion_r1618729365) → [changes: aac8e1c](https://github.com/ethereum/yellowpaper/pull/906/commits/aac8e1c55acb03242b0d1e7123298362e087fba3)

<details>
<summary>View Screenshot Diff</summary>

|Before|After|
|-|-|
|4895_11_before|4895_11_after|
|4895_0_before|4895_0_after4895_0_after_1|
|4895_1_before|4895_1_after|
|4895_2_before|4895_2_after|
|4895_3_before|4895_3_after|
|4895_4_before|4895_4_after|
|4895_5_before|4895_5_after|
|4895_6_before|4895_6_after|
|4895_7_before|4895_7_after|
|4895_8_before|4895_8_after|
|4895_9_before|4895_9_after_1|
|4895_10_before|4895_10_after|

</details>

# EIP-6049: Deprecate SELFDESTRUCT
- [specs](https://eips.ethereum.org/EIPS/eip-6049)
- [commit changes](https://github.com/ethereum/yellowpaper/pull/906/commits/20ebaa7cb3947992b0c2a5bbae4881036f593372)
- [review](https://github.com/ethereum/yellowpaper/pull/906#discussion_r1618785791) → [changes: 22c2ca5](https://github.com/ethereum/yellowpaper/pull/906/commits/22c2ca5c9b56e180dfb5ae27b6b57b31daaa1ee2)

<details>
<summary>View Screenshot Diff</summary>

|Before|After|
|-|-|
|6049_0_before|6049_0_after_1|

</details>

# General Updates
- [commit changes](https://github.com/ethereum/yellowpaper/pull/906/commits/5df35cf27feb0b7d5397db8f8592f0195868e93f)
- [review](https://github.com/ethereum/yellowpaper/pull/906#discussion_r1618795416) → [changes: 878ee5b](https://github.com/ethereum/yellowpaper/pull/906/commits/878ee5bd9977db043da8a293fe510737ee3be1d5)

<details>
<summary>View Screenshot Diff</summary>

|Before|After|
|-|-|
|shanghai_0_before|shanghai_0_after|
|shanghai_1_before|shanghai_1_after|
|shanghai_2_before|shanghai_2_after|
|shanghai_3_before|shanghai_3_after_1|

</details>

### CODE REVIEW

**Comment by pldespaigne** [comment]
Bump @yperbasis @timbeiko
Let me know, if you need more detail on the PR changes, or anything else!

**Comment by yperbasis** [comment]
> Bump @yperbasis @timbeiko Let me know, if you need more detail on the PR changes, or anything else!

Currently I don't have the capacity to review this PR unfortunately :(

**Comment by pldespaigne** [comment]
@yperbasis 
I understand, do you know of anyone that could have a look at this?
Should I try to bring some non-maintainer peoples in the convo to give feedback on my changes?
I've signed a grant agreement with the EF to update the YP to Shanghai & Cancun, that's why I'll try to help you as much as I can to merge this.

**Comment by alexprut** [comment]
@pldespaigne not a maintainer, nor fully up to date with all the protocol changes, contributed (small) changes years ago.

However, the work you are doing is interesting, if it's helpful I can sped a spike next week to review and provide feedback.

**Comment by pldespaigne** [comment]
That would be awesome 🙏
Anyone's feedback would be great!

File: Paper.tex
Comment (alexprut) [inline Paper.tex:888]: This part LGTM.

Small observations:
1) Are the curly brackets necessary? (i.e. can `\{H_c\}` just be  `H_c`). I guess because the union operation is defined on sets, if that is the criteria then formula `(74)` on `a ∪ T_t` is missing the brackets.
2) My understanding is that this EIP is retro-compatible, still should it specify in any comment/footnote that it was introduced with `EIP-3651`?
3) (_Out of scope of the PR_) While reading the Block definition section 4.3, `H_c` could also state the following: "(also known as the coinbase address)"

File: Paper.tex
Comment (alexprut) [inline Paper.tex:2200]: Changes for `EIP-3855` LGTM.

- Not fully sure thought if formula nr `(157)` also needs a change (haven't spend much time on it)?

File: Paper.tex
Comment (alexprut) [inline Paper.tex:795]: Since this `EIP-3860` required a "network update", should it indicate that this was introduced with `EIP-3860`?

**Comment by alexprut** [comment]
Soundness of `EIP-3860` changes LGTM.

File: Paper.tex
Comment (pldespaigne) [reply Paper.tex:888]: Re: EIP-3651
1. Looking at other formulas of the YP, you are right, I just removed `{ }` for clarity
2. I don't think so, EIPs are rarely cited except for major changes such as EIP1559.
3. Totally agree with you, I didn't added it though, cause it's out of scope and I don't want to add more noise, this PR will already be hard to merge, might be worth it to add it as a tiny separate PR.

Comment (pldespaigne) [reply Paper.tex]: Changes in [689eb33](https://github.com/ethereum/yellowpaper/pull/906/commits/689eb3353be76717d1378640b06551d73826d6a4)

File: Paper.tex
Comment (pldespaigne) [reply Paper.tex:2200]: Not sure what formula you are talking about (the numbers changes almost with every commit).
Assuming you are talking about this formula:
image
If so, I don't think it need update as it only refer to $w  \in [PUSH1, PUSH32]$ and `PUSH0` is not in that range

File: Paper.tex
Comment (pldespaigne) [reply Paper.tex:2111]: [EIP-3860](https://eips.ethereum.org/EIPS/eip-3860#parameters) define it as its own `INITCODE_WORD_COST` constant, also I would find it confusing to use $G_{base}$ because of the description: "Amount of gas to pay for operations of the set $W_{base}$.". What do you think?

In another subject, your comment made me realize that function $N(x)$ was already used as pointed in your review of `EIP-3855` above, I'll change it to $R(x)$ which I think is not used.

File: Paper.tex
Comment (pldespaigne) [reply Paper.tex:795]: I'm not in favor as it's not a "major" change, but I'm open to discussion.

File: Paper.tex
Comment (pldespaigne) [reply Paper.tex:2696]: You are right, EIP defines it as it's own constant, but since the YP used the same formalism on post deploy code check (Section 7), I decided to keep it like this.

image

File: Paper.tex
Comment (pldespaigne) [reply Paper.tex:2111]: Changes in [#5a342a7](https://github.com/ethereum/yellowpaper/pull/906/commits/5a342a742d1c6c04b23b143a7c49b381b7a85660)

**Comment by pldespaigne** [comment]
Thanks a lot for your feedbacks @alexprut 🙏

**Comment by pldespaigne** [comment]
@nicksavers @pirapira Any time on your side to have a look at this?
Or maybe you could delegate that to someone?

**Comment by alexprut** [comment]
> Thanks a lot for your feedbacks @alexprut 🙏

More than happy to be helpful, thanks for updating the YP.
@pldespaigne FYI I still need to look at and review commits `EIP-4895`, `EIP-6049` and `General Updates`, aiming at finishing the review by EOW.

File: Paper.tex
Comment (alexprut) [reply Paper.tex:2200]: Sounds great, indeed was referencing to the formula you posted.

File: Paper.tex
Comment (alexprut) [reply Paper.tex:2111]: > [EIP-3860](https://eips.ethereum.org/EIPS/eip-3860#parameters) define it as its own `INITCODE_WORD_COST` constant, also I would find it confusing to use Gbase because of the description: "Amount of gas to pay for operations of the set Wbase.". What do you think?
> 
> In another subject, your comment made me realize that function N(x) was already used as pointed in your review of `EIP-3855` above, I'll change it to R(x) which I think is not used.

Your call, don't have strong opinions, it can always be simplified post this PR.

File: Paper.tex
Comment (alexprut) [reply Paper.tex]: > I'm not in favor as it's not a "major" change, but I'm open to discussion.

Sounds good, don't have strong opinions on this one.

File: Paper.tex
Comment (alexprut) [inline Paper.tex:426]: `Gwei` could just go as a definition in section `2.1.`

File: Paper.tex
Comment (alexprut) [inline Paper.tex:418]: FYI, not a big issue, `W` is already defined in the paper as `W(w, μ)`

File: Paper.tex
Comment (alexprut) [inline Paper.tex:1411]: 3 stages now

File: Paper.tex
Comment (alexprut) [inline Paper.tex:1414]: My understanding is that `validate transactions` precede  `executing withdrawals`

File: Paper.tex
Comment (alexprut) [inline Paper.tex:1421]: FYI, not necessarily an issue, function `P` is already defined in the paper.

File: Paper.tex
Comment (alexprut) [inline Paper.tex:2797]: Suggesting adding that it may be subject "to **breaking** changes"

File: Paper.tex
Comment (alexprut) [inline Paper.tex:1389]: proposed **a the** block

File: Paper.tex
Comment (alexprut) [inline Paper.tex:206]: Out of scope of this PR, was thinking if going forward there should be another column besides `Block Number`, e.g. `Timestamp` or whatever will be the canonical trigger.

**Review by alexprut — APPROVED** [review]

File: Paper.tex
Comment (pldespaigne) [reply Paper.tex:426]: Good catch! Since `EIP-4895` explicitly define this value as being denominated in `Gwei`, I think you are right.

File: Paper.tex
Comment (pldespaigne) [reply Paper.tex:418]: We don't have enough letters in greek & latin alphabet 😅
As you said, I think it's not a big issue, `W` for **W**ithdrawal is easy to remember, it follows the same logic as `T` for **T**ransaction, etc. Also, there is already `H` for block **H**eader even if the normal halting function is defined as `H(μ, I)`

File: Paper.tex
Comment (pldespaigne) [reply Paper.tex:1414]: That's a good question! The EIP only states that it should happen after transaction *execution* but it says nothing about where it should happen during block finalization.
The transaction *validation* step purpose in the YP is simply meant to set the block header's gas used value ($B_{Hg}$), since withdrawals incur no gas cost, it has no impact if its run before or after withdrawal execution.

File: Paper.tex
Comment (pldespaigne) [reply Paper.tex:427]: changes in [55a4099](https://github.com/ethereum/yellowpaper/pull/906/commits/55a4099d6366a2425e4c7c8ad1ba2687cd1abc89)

File: Paper.tex
Comment (pldespaigne) [reply Paper.tex]: changes in [55a4099](https://github.com/ethereum/yellowpaper/pull/906/commits/55a4099d6366a2425e4c7c8ad1ba2687cd1abc89)

File: Paper.tex
Comment (pldespaigne) [reply Paper.tex]: changes in [55a4099](https://github.com/ethereum/yellowpaper/pull/906/commits/55a4099d6366a2425e4c7c8ad1ba2687cd1abc89)

File: Paper.tex
Comment (pldespaigne) [reply Paper.tex]: changes in [91f0547](https://github.com/ethereum/yellowpaper/pull/906/commits/91f054787b4f38d15bddcefa09bc98ea0a62144b)

File: Paper.tex
Comment (pldespaigne) [reply Paper.tex:207]: That's a good idea imo!

File: Paper.tex
Comment (pldespaigne) [reply Paper.tex]: changes in [2b98d68](https://github.com/ethereum/yellowpaper/pull/906/commits/2b98d68e5ea35e553fcc4fdab9c122ca91f55f38)

**Comment by yperbasis** [comment]
> Thanks again for your time @alexprut 🙏 I've made all the requested changes and update the PR description. I'll now try to bring more eyes on this discussion!
> 
> @yperbasis how many approvals would be required for you to be comfortable in merging this PR?

Sorry guys, I was on holiday and next week I'll be at a workshop. I'll take a look during the week starting from 24 June.

**Review by yperbasis — COMMENTED** [review]
Please also update BRANCHES.md & README.md (see how it was done in PR #894 for Paris).

**Comment by yperbasis** [comment]
Meanwhile I've created branch `paris` from the current `master`.

**Comment by pldespaigne** [comment]
I've rebased the PR with latest changes from `master`.
I've also updated `README.md` & `BRANCHES.md` as requested.

Let me know if I can do anything else!

**Review by yperbasis — APPROVED** [review]

**Comment by pldespaigne** [comment]
Thanks a lot for your time @yperbasis!
I'll now start to work on the Cancun update, should I open a single PR for the whole update, or do you prefer to review it eip by eip?

**Comment by yperbasis** [comment]
> Thanks a lot for your time @yperbasis! I'll now start to work on the Cancun update, should I open a single PR for the whole update, or do you prefer to review it eip by eip?

Thank you! It's easier for me to review eip by eip.

**Comment by pldespaigne** [comment]
Alright, I'll do it eip by eip then!
Also it looks like the merge of this PR didn't triggered a deploy of the new version of the pdf: https://ethereum.github.io/yellowpaper/paper.pdf

**Comment by pldespaigne** [comment]
Thanks a lot 🙏
Will do, for sure!

**Comment by yperbasis** [comment]
> Alright, I'll do it eip by eip then! Also it looks like the merge of this PR didn't triggered a deploy of the new version of the pdf: https://ethereum.github.io/yellowpaper/paper.pdf

I've re-run the job and https://ethereum.github.io/yellowpaper/paper.pdf seems to be fine now.

**Pull request merged**
